### ISSUE TITLE
Something is wrong with DB of my instance

### ISSUE THREAD

**Issue by noraworld** [tootsuite/mastodon#4856]
Recently I have changed the environment of my instance ([mastodon.noraworld.jp](https://mastodon.noraworld.jp)) from Docker to non-Docker. The DB takes some strange behaviors since then.

@clworld, @unarist and @akihikodaki maybe know what happened in my DB well I think. Could you give me a hand?

## 1. Split into two accounts
| id = 1441 | id = 7643 |
| --- | --- |
| ![clworld_before](https://user-images.githubusercontent.com/17060801/30224208-d07c9014-9508-11e7-8144-b886fd3a6f2c.png) | ![clworld_after](https://user-images.githubusercontent.com/17060801/30224274-12880d30-9509-11e7-8d7e-7ec9b7d5baf5.png) |

Some users split into two accounts, which means one same user has two account records in the DB.

```
psql> SELECT username, domain FROM accounts GROUP BY username, domain HAVING count(*) >= 2;

(38 rows)
```

The result of this query should be `0 rows` because multiple records with same username and same domain cannot be created by unique constraint. But somehow created in my DB.

I dumped this DB and restored the dump file to test environment. Strangely, the duplicated account records were inserted into DB in spite of unique constraint violation.

## 2. Mention not reached
![mention_not_reached](https://user-images.githubusercontent.com/17060801/30224337-53c78a1e-9509-11e7-9fe4-0b9bfd1bd707.png)

I cannot send a mention to some users. The error "422 Validation failed: Username has already been taken" occurs.

I cannot send it to not all users, but some users. And also I was properly able to send a mention to the user whom I cannot send it now when working on Docker.

## 3. Cannot search toot
![cannot_search_toot](https://user-images.githubusercontent.com/17060801/30224342-59e2e6dc-9509-11e7-9947-ad0b5eb1edf4.png)

The error message is same as the previous problem. Probably No. 3 cause is same as No. 2 cause.

This user is not shown on timelines though I follow him.

***

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by noraworld** [comment]
The version of PostgreSQL is 9.6 on both Docker and non-Docker. I think the problem with compatibility is not related to the cause.

I have just copied `mastodon/postgres` and `mastodon/redis` to `/var/lib/pgsql/9.6/data` and `/var/lib/redis` when I have changed the environment from Docker to non-Docker, not dumped. Is there any problem with this way of DB restore?

**Comment by unarist** [comment]
From the previous discussion, some indexes seems to be corrupted. At least, we have confirmed `index_accounts_on_uri` and `index_accounts_on_username_and_domain` have missed some records, then they can't find some existing record and also can violate unique indexes.

> 422 Validation failed: Username has already been taken

This error will be raised if you violates the unique index for (username, domain) on the accounts table. Probably that user cannot be found by indexes for uri column, but that unique index still covers it.

**Comment by Gargron** [comment]
I believe pg_dump and pg_restore are the safest way to move data between postgres servers. It's likely the file system was in an incomplete state when you copied it (pg keeps some stuff in memory) 

What you'll need to do is rebuild the indices. I think there is a sql command for that. But first you'll have to manually resolve unique conflicts, i.e. delete duplicate records

**Comment by noraworld** [comment]
Thank you for your help. It seemed the cause of this fault was the whole PostgreSQL database files were copied to migrate the environment. I didn't know which account of splitted accounts should be deleted because the toots were coming from each account of them. So the reindex operation could not be executed in the end.

I reverted the database to the previous one that have worked completely. Then I backed up it and restored it to the non-Docker environment by running `pg_dump` and `pg_restore` as @Gargron says. My current instance works properly now.

I close this issue, but I might ask for your help if such a strange fault occurred again. Thank you.
